definitions/npm/firebase_v4.x.x/flow_v0.46.x-/firebase_v4.x.x.js

```diff
@@ -0,0 +1,410 @@
+  delete(): Promise<void>;
+  getIdToken(forceRefresh?: boolean): Promise<string>;
+  getToken(forceRefresh?: boolean): Promise<string>;
+  linkAndRetrieveDataWithCredential(credential: AuthCredential): Promise<UserCredential>;
+  linkWithCredential(credential: AuthCredential): Promise<User>;
+  linkWithPhoneNumber(phoneNumber: string, applicationVerifier: ApplicationVerifier): Promise<ConfirmationResult>;
+  linkWithPopup(provider: OAuthProvider): Promise<UserCredential>;
+  reauthenticateAndRetrieveDataWithCredential(credential: AuthCredential): Promise<void>;
+  reauthenticateWithCredential(credential: AuthCredential): Promise<void>;
+  reauthenticateWithPhoneNumber(phoneNumber: string, applicationVerifier: ApplicationVerifier): Promise<ConfirmationResult>;
+  reload(): Promise<void>;
+  sendEmailVerification(): Promise<void>;
+  toJSON(): Object;
+  unlink(providerId: string): Promise<User>;
+  updateEmail(newEmail: string): Promise<void>;
+  updatePassword(newPassword: string): Promise<void>;
+  updatePhoneNumber(phoneCredential: AuthCredential): Promise<void>;
+  updateProfile(profile: { displayName: ?string, photoURL: ?string }): Promise<void>;
+}
+
+declare class EmailAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  providerId: string;
+  credential(email: string, password: string): AuthCredential;
+}
+
+declare class FacebookAuthProvider extends AuthProvider  {
+  PROVIDER_ID: string;
+  credential(token: string): AuthCredential;
+  addScope(scope: string): FacebookAuthProvider;
+  setCustomParameters(customOAuthParameters: Object): FacebookAuthProvider;
+}
+
+declare class GithubAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  credential(token: string): AuthCredential;
+  addScope(scope: string): GithubAuthProvider;
+  setCustomParameters(customOAuthParameters: Object): GithubAuthProvider;
+}
+
+declare class GoogleAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  credential(idToken?: string, accessToken?: string): AuthCredential;
+  addScope(scope: string): GoogleAuthProvider;
+  setCustomParameters(customOAuthParameters: Object): GoogleAuthProvider;
+}
+
+declare class PhoneAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  constructor(auth?: Auth): PhoneAuthProvider;
+  credential(verificationId: string, verificationCode: string): AuthCredential;
+  verifyPhoneNumber(phoneNumber: string, applicationVerifier: ApplicationVerifier): Promise<string>;
+}
+
+declare class TwitterAuthProvider extends AuthProvider {
+  PROVIDER_ID: string;
+  credential(token: string, secret: string): AuthCredential;
+  setCustomParameters(customOAuthParameters: Object): this;
+}
+
+declare type OAuthProvider = | FacebookAuthProvider
+                             | GithubAuthProvider
+                             | GoogleAuthProvider
+                             | TwitterAuthProvider;
+
+/****** database ******/
+declare type FirebaseValue = {} | Array<FirebaseValue> | string | number | boolean | null;
```

Comment:
Your typing for FirebaseValue is different from the def. See:
- [Type def for update](https://github.com/stevenyap/flow-typed/blob/1c3af2f28cea3333ed6b192f6117b63e632918fb/definitions/npm/firebase_v4.x.x/flow_v0.34.x-v0.45.x/firebase_v4.x.x.js#L370)
- [Type def for FirebaseValue](https://github.com/stevenyap/flow-typed/blob/1c3af2f28cea3333ed6b192f6117b63e632918fb/definitions/npm/firebase_v4.x.x/flow_v0.34.x-v0.45.x/firebase_v4.x.x.js#L251)

Anyhow, I have recreated your code and found that the update throws a flow error.
After digging around, I realise the issue is with the variance typing of the object.

In short, when we write `update(wb)`, flow requires that the wb be an invariant of `{[path:string]: FirebaseValue}` and hence throws an error. The correct way is to declare update's param as a covariant which is `{+[path: string]: FirebaseValue}` (Note the extra `+`).

Please see my fix for your code:
[Flow Try](https://flow.org/try/#0PTAEAEDMBsHsHcBQJQHkBGArApgYwC6j4CeADtqAJYDOVAdgG4CGATpU3foieaAOqwWAa3SxYQ0AF5QAb1C5YdOnnyVFALlmhoTdNmibq+NnQDmoAL6XE3MhQBilFtnRNq2AGpNoAVwqTEUFAAH1kLQJDQAEEWFiZiAB5HZ1d3L19sAD4I0KMTUxzQOh8AWz0WQtFYaGwOQuLoaBsUAGVYUB9SABMmVTNQDlBYLBUiO1BqAAtYH2gu0D1QEtYhbHm3eVhmNg4uSB86AjU6Du7e7AAKZgzqTTkAagBtUl7Jw2NKMwBdTWSXN083j8lgAlJoGLBKPMZBEFHRqNVsAA6OCmK5A7DUEGIcKIOFGUDwdCaATCKoSaRyOHKI4aLQ6PQGUAAcnwmPwzMsuM6PTZFyJIKAA)

I have updated PR #955 with the fix above.